### PR TITLE
Categorical.(get|from)_dummies

### LINKED ISSUE #8745: API/ENH: from_dummies
Motivating from [SO](http://stackoverflow.com/questions/26762100/reconstruct-a-categorical-variable-from-dummies-in-pandas)

This is the inverse of `pd.get_dummies`. So maybe `invert_dummies` is better?
I think this name makes more sense though. 

This seems a reasonable way to do it. Am I missing anything?

```
In [46]: s = Series(list('aaabbbccddefgh')).astype('category')

In [47]: s
Out[47]: 
0     a
1     a
2     a
3     b
4     b
5     b
6     c
7     c
8     d
9     d
10    e
11    f
12    g
13    h
dtype: category
Categories (8, object): [a < b < c < d < e < f < g < h]

In [48]: df = pd.get_dummies(s)

In [49]: df
Out[49]: 
    a  b  c  d  e  f  g  h
0   1  0  0  0  0  0  0  0
1   1  0  0  0  0  0  0  0
2   1  0  0  0  0  0  0  0
3   0  1  0  0  0  0  0  0
4   0  1  0  0  0  0  0  0
5   0  1  0  0  0  0  0  0
6   0  0  1  0  0  0  0  0
7   0  0  1  0  0  0  0  0
8   0  0  0  1  0  0  0  0
9   0  0  0  1  0  0  0  0
10  0  0  0  0  1  0  0  0
11  0  0  0  0  0  1  0  0
12  0  0  0  0  0  0  1  0
13  0  0  0  0  0  0  0  1

In [50]: x = df.stack()

# I don't think you actually need to specify ALL of the categories here, as by definition
# they are in the dummy matrix to start (and hence the column index)
In [51]: Series(pd.Categorical(x[x!=0].index.get_level_values(1)))
Out[51]: 
0     a
1     a
2     a
3     b
4     b
5     b
6     c
7     c
8     d
9     d
10    e
11    f
12    g
13    h
Name: level_1, dtype: category
Categories (8, object): [a < b < c < d < e < f < g < h]
```

NB. this is buggy ATM.

```
In [51]: Series(pd.Categorical(x[x!=0].index.get_level_values(1)),categories=df.categories)
```

### PR DESCRIPTION
**Pull request by clbarnes**
Simply converting categorical variables to and from dummy variables.

- [x] closes #8745
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Intentionally smaller-scoped than #31795 (and indeed `get_dummies`) as a broadly useful MVP which can be chained with other basic functionality. The tests are fairly rudimentary and I welcome any edge cases which should be picked out.

### from_dummies

* class method rather than free function: Keeps categorical-related functionality together, reduces surface in the global namespace, more obvious what is produced.
* silently drop column with NA header: wasn't sure about this. Maybe it should raise a warning?
* No handling of masked dataframes or dataframes with NA values
* No subsetting or renaming of columns: callers can do this themselves

### to_dummies

* Name: I went for `Categorical.to_dummies` instead of matching the free function `get_dummies`. The symmetry of to/from aids understanding, and using `get_` might imply A) that something is being got, which it isn't or B) signature/ feature parity with the existing method, which wasn't a design goal for me.
* `to_dummies` return type: `cls.to_dummies` returns bools, where `get_dummies` returns uint8s by default, which doesn't make a lot of sense to me as we are representing boolean data (and they're the same in memory anyway). Dummy variables are generally used for regression where a continuous variable is required, so ints don't get us any closer to what we may want, and being able to index into the categories using the row may be valuable.
* No `dtype` argument: I didn't see any benefit of `cls.to_dummies(dtype=float)` over `cls.to_dummies().astype(float)`. The latter is more explicit, no slower, and minimises API surface.
* No `prefix`, `prefix_sep` arguments: these unnecessarily assume string column headers. If someone wants to rename their columns, they can: IMO it's not a core requirement of this method.
* `dumma_na` replaced with `na_column`: if we're including the argument, we may as well let the user decide what they want to call their column, using the dtype they prefer (e.g. `"other"`, `-1` etc). They can always supply `np.nan` if they want `get_dummies`-like behaviour.
* No `sparse` argument: This I regret. Producing a sparse array would be valuable. However, it would drastically complicate the method, so I left it out for the MVP. The caller can always sparsify it after it's produced, so long as RAM isn't an issue for the temporary df.
* No `drop_first` argument: If someone wants to drop one of their columns, they can (`cls.to_dummies().drop(columns="my_col")`): again, not a core requirement of this method. Broadly speaking, I'm not in favour of adding arguments to save the caller <=1 line of their own code unless there are e.g. speed gains.

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -379,6 +382,110 @@ def __init__(
...
+        --------
+        >>> df = pd.DataFrame(
+        ... [[1, 0, 0], [0, 1, 0], [0, 0, 1]], columns=["a", "b", "c"]
+        ... )
+        >>> Categorical.from_dummies(df)
+        [a, b, c]
+        Categories (3, object): [a, b, c]
+        """
+        # GH 8745
+        from pandas import Series
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py:421]: Is this only used to get `index_into`, which is only used to get `len(index_into)`? If so, why not just have
```python
mult_by = np.arange(1, df.shape[1]+1)
```?
?

```

File: pandas/core/arrays/categorical.py [hunk 2]
```diff
@@ -379,6 +382,110 @@ def __init__(
         self._dtype = self._dtype.update_dtype(dtype)
         self._codes = coerce_indexer_dtype(codes, dtype.categories)
 
+    @classmethod
+    def from_dummies(cls, dummies: "DataFrame", ordered=None):
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py:386]: Can you annotate `ordered`, as well as the return type?

File: pandas/core/arrays/categorical.py [hunk 3]
```diff
@@ -379,6 +382,110 @@ def __init__(
...
+        """
+        Create a `Categorical` using a ``DataFrame`` encoding those categories
+        as dummy/ one-hot encoded variables.
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py:389]: The [docstring guide](https://pandas.pydata.org/docs/development/contributing_docstring.html) asks that the short summary fit on a single line, is there a way to reword it?

File: pandas/core/arrays/categorical.py [hunk 4]
```diff
@@ -379,6 +382,110 @@ def __init__(
...
+        df = dummies.drop(columns=np.nan, errors="ignore").astype(bool)
+
+        if (df.sum(axis=1) > 1).any():
+            raise ValueError("Some rows belong to >1 category")
+
+        index_into = Series([np.nan] + list(df.columns))
+        mult_by = np.arange(1, len(index_into))
+
+        codes = (df.astype(int) * mult_by).sum(axis=1) - 1
+        codes[codes.isna()] = -1
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py:432]: I may have misunderstood what's happening here, but would it work to use `pd.factorize`?

File: pandas/tests/arrays/categorical/test_api.py [hunk 5]
```diff
@@ -399,6 +399,19 @@ def test_remove_unused_categories(self):
...
+    @pytest.mark.parametrize(
+        "vals",
+        [
+            ["a", "b", "b", "a"],
+            ["a", "b", "b", "a", np.nan],
+            [1, 1.5, "a", (1, "b")],
+            [1, 1.5, "a", (1, "b"), np.nan],
+        ],
+    )
+    def test_to_dummies(self, vals):
```
Comment (MarcoGorelli) [inline pandas/tests/arrays/categorical/test_api.py:411]: I think the issue number as a comment should go in every new test

File: pandas/core/arrays/categorical.py [hunk 6]
```diff
@@ -379,6 +382,110 @@ def __init__(
...
+        The ``DataFrame`` must be coercible to boolean,
+        and have no more than one truthy value per row.
+        The columns of the ``DataFrame`` become the categories of the `Categorical`.
+        A column whose header is NA will be dropped.
+
+        Parameters
+        ----------
+            dummies : DataFrame of bool-like
+            ordered : bool
+                Whether or not this Categorical is ordered.
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py:400]: Move these back one indent

File: doc/source/whatsnew/v1.1.0.rst [hunk 7]
```diff
@@ -255,6 +255,7 @@ Other enhancements
 - :meth:`HDFStore.put` now accepts `track_times` parameter. Parameter is passed to ``create_table`` method of ``PyTables`` (:issue:`32682`).
 - Make :class:`pandas.core.window.Rolling` and :class:`pandas.core.window.Expanding` iterable（:issue:`11704`)
 - Make ``option_context`` a :class:`contextlib.ContextDecorator`, which allows it to be used as a decorator over an entire function (:issue:`34253`).
+- :class:`~pandas.core.arrays.categorical.Categorical` now has methods for converting to and from dummy/ one-hot encoded variables: :meth:`Categorical.to_dummies` and :meth:`Categorical.from_dummies` respectively. :meth:`Categorical.to_dummies` is smaller in scope than :func:`~pandas.core.reshape.reshape.get_dummies`, which can still be used if you require the extra flexibility.
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v1.1.0.rst:258]: Put the issue number at the end

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Cool, thanks @clbarnes !

File: pandas/core/arrays/categorical.py (hunk 2 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:386]: Is there a preference for how to annotate the self return type? As I understand it, you can do it with a `TypeVar`, with a string, or by using `from __future__ import annotations` (although only in 3.7+).

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:421]: My bad, that was a hangover from before I started using `from_codes`.

File: pandas/core/arrays/categorical.py (hunk 4 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:430]: I may be missing what `factorize` does, but I don't think it would save any effort here -  factorize wants a 1D array (which this isn't until `codes` is instantiated), and then doesn't produce a `Categorical` unless given one, so the last line wouldn't change either.

File: pandas/core/arrays/categorical.py [hunk 8]
```diff
@@ -379,6 +382,107 @@ def __init__(
...
+        Examples
+        --------
+        >>> df = pd.DataFrame(
+        ... [[1, 0, 0], [0, 1, 0], [0, 0, 1]], columns=["a", "b", "c"]
+        ... )
+        >>> Categorical.from_dummies(df)
+        [a, b, c]
+        Categories (3, object): [a, b, c]
+        """
+        # GH 8745
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py:421]: I don't think the issue number needs to be here, should only be necessary in the tests

File: pandas/core/arrays/categorical.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/core/arrays/categorical.py:430]: OK yes, I was thinking something like
```python
codes, _ = factorize(df.idxmax(axis=1))
```
, but then it'd be necessary to deal with the case when there's a row without any ones, which'd require more code and would be no shorter/clearer than what you've already got. So, ignore the `factorize` suggestion :)

Could of thiings then:
- is `astype(int)` necessary? It should work without it
- perhaps there would be a comment explaining why you do `-1` at the end of this line
```python
        codes = (df.astype(int) * mult_by).sum(axis=1) - 1
```
?
Now that I've run the code, it's obvious what it does, but it wasn't when I was just reading it

**Review by jreback — CHANGES_REQUESTED** [review]
will look soon

**Comment by jreback** [comment]
likely would want a doc-example of both of these near in https://pandas.pydata.org/docs/dev/user_guide/reshaping.html#computing-indicator-dummy-variables and/or links to categorical.rst

File: pandas/core/arrays/categorical.py (hunk 4 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py]: You're right, the `astype` wasn't necessary. There's a comment now, best way I could think of explaining it but I can go for prose if you prefer.

File: pandas/core/arrays/categorical.py [hunk 9]
```diff
@@ -379,6 +382,109 @@ def __init__(
...
+           a      b      other
+        0  True   False  False
+        1  False  True   False
+        2  False  False  True
+
+        See Also
+        --------
+        :func:`pandas.get_dummies`
+        """
+        from pandas import DataFrame, CategoricalIndex, Series
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py:477]: Why is this not using the existing `get_dummies` implementation? This shold be identical to `pd.get_dummies(Categorical)`, right?

File: pandas/core/arrays/categorical.py [hunk 10]
```diff
@@ -379,6 +382,109 @@ def __init__(
...
+        mult_by = np.arange(df.shape[1]) + 1
+        #  000            000    0   -1
+        #  010            020    2    1
+        #  001 * 1,2,3 => 003 -> 3 -> 2 = correct codes
+        #  100            100    1    0
+        codes = (df * mult_by).sum(axis=1) - 1
+        codes[codes.isna()] = -1
+        return cls.from_codes(codes, df.columns.values, ordered=ordered)
+
+    def to_dummies(self, na_column=None) -> "DataFrame":
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py:435]: I think this should be called `get_dummies` I don't think the justification for deviating from that name is strong enough to warrant a different name.

I also think it should exactly match the signature of `get_dummies`.

File: pandas/core/arrays/categorical.py [hunk 11]
```diff
@@ -379,6 +382,109 @@ def __init__(
...
+    def from_dummies(
+        cls, dummies: "DataFrame", ordered: Optional[bool] = None
+    ) -> "Categorical":
+        """Create a `Categorical` using a ``DataFrame`` of dummy variables.
+
+        The ``DataFrame`` must be coercible to boolean,
+        and have no more than one truthy value per row.
+        The columns of the ``DataFrame`` become the categories of the `Categorical`.
+        A column whose header is NA will be dropped;
+        any row with a NA value will be uncategorised.
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py:395]: Need examples showing these.

File: pandas/core/arrays/categorical.py [hunk 12]
```diff
@@ -379,6 +382,109 @@ def __init__(
...
+
+        Parameters
+        ----------
+        dummies : DataFrame of bool-like
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py:399]: Just the type on the first line. The structure can go on the second line.

File: pandas/core/arrays/categorical.py [hunk 13]
```diff
@@ -379,6 +382,109 @@ def __init__(
...
+        ordered : bool
+            Whether or not this Categorical is ordered.
+
+        Raises
+        ------
+        ValueError
+            If a sample belongs to >1 category
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py:406]: Add an example.

File: pandas/tests/arrays/categorical/test_constructors.py [hunk 14]
```diff
@@ -643,3 +645,55 @@ def test_constructor_string_and_tuples(self):
...
+        raw = ["a", "a", "b", "c", "c", "a", np.nan]
+        dummies = get_dummies(raw)
+        cats = Categorical.from_dummies(dummies)
+        assert list(cats)[:-1] == raw[:-1]
+        assert pd.isna(list(cats)[-1])
+
+    def test_from_dummies_gt1(self):
+        # GH 8745
+        dummies = DataFrame([[1, 0, 1], [0, 1, 0], [0, 0, 1]], columns=["a", "b", "c"])
+        with pytest.raises(ValueError):
```
Comment (TomAugspurger) [inline pandas/tests/arrays/categorical/test_constructors.py:667]: Make sure we're raising an informative error message here.

File: pandas/tests/arrays/categorical/test_constructors.py [hunk 15]
```diff
@@ -643,3 +645,55 @@ def test_constructor_string_and_tuples(self):
...
+    def test_from_dummies(self):
+        # GH 8745
+        raw = ["a", "a", "b", "c", "c", "a"]
+        dummies = get_dummies(raw)
+        cats = Categorical.from_dummies(dummies)
+        assert list(cats) == raw
+
+    def test_from_dummies_nan(self):
+        # GH 8745
+        raw = ["a", "a", "b", "c", "c", "a", np.nan]
```
Comment (TomAugspurger) [inline pandas/tests/arrays/categorical/test_constructors.py:658]: Is `nan` the only supported NA value that's dropped?

File: pandas/tests/arrays/categorical/test_constructors.py [hunk 16]
```diff
@@ -643,3 +645,55 @@ def test_constructor_string_and_tuples(self):
         c = pd.Categorical(np.array(["c", ("a", "b"), ("b", "a"), "c"], dtype=object))
         expected_index = pd.Index([("a", "b"), ("b", "a"), "c"])
         assert c.categories.equals(expected_index)
+
+    def test_from_dummies(self):
```
Comment (TomAugspurger) [inline pandas/tests/arrays/categorical/test_constructors.py:649]: Can you parametrize this over sparse = True/False?

Comment (clbarnes) [reply pandas/tests/arrays/categorical/test_constructors.py:649]: For the sake of simplicity I have not made an attempt to support loading from a sparse array but I can give it a go!

**Review by clbarnes — COMMENTED** [review]
Thanks for your comments @TomAugspurger ! I've addressed most of them but think the `to_dummies` and `get_dummies` shared functionality is probably worth discussion on the main thread.

**Comment by clbarnes** [comment]
> I think this should be called get_dummies I don't think the justification for deviating from that name is strong enough to warrant a different name.
>
> I also think it should exactly match the signature of get_dummies.

@TomAugspurger My initial implementation was a thin wrapper around `get_dummies` but given how simple the "raw" implementation is I didn't see a huge benefit to it. I could switch back, although I had wondered whether it might be better for for `get_dummies` to actually call `to_dummies` in some cases to avoid some logic.

I do feel more strongly about keeping the name and API different to `get_dummies`, though. `get_` methods imply retrieving something from the object rather than converting it to a different form. `from_` is IMO the most obvious name for the alternate constructor, and if `from_` exists, the opposite ought to be `to_` (`as_` or `into_` would also be OK but I feel they imply more mutation/ destructive operations). I personally have a minor nit with the naming of pandas' IO methods on this basis - `read_csv` and `to_csv` are from two different opposite pairs, which increased the time it took for me to remember them.

My justifications for not supporting all of the `get_dummies` interface are in the PR description. Supporting the whole `get_dummies` API in `to_dummies` would necessitate supporting all of the reverse operations in `from_dummies` (which is the key part of this PR; `to_dummies` is more of an afterthought), which I considered to be a large complexity overhead for not a lot of gain in utility. If we're not supporting the whole API, they shouldn't have the same name.

**Comment by TomAugspurger** [comment]
> I do feel more strongly about keeping the name and API different to get_dummies

Yeah, your reasons for `to_dummies` all make sense. But in my mind it comes down to a values judgement between

1. Symmetry the two methods, which favors `to_dummies` over `get_dummies`, as `to_dummies` aligns better with `from_dummies`
2. Consistency with the established name `get_dummies`.

In my opinion, 2 wins out.

> Supporting the whole get_dummies API in to_dummies would necessitate supporting all of the reverse operations in from_dummies

That would indeed be helpful. Which parameters / behaviors are especially hard to support? I'm happy to raise a `NotImplementedError` for things that could be done in a followup PR if anyone is interested.

File: doc/source/user_guide/categorical.rst [hunk 17]
```diff
@@ -129,6 +129,48 @@ This conversion is likewise done column by column:
     df_cat['A']
     df_cat['B']
 
+Dummy / indicator / one-hot encoded variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some operations, like regression and classification,
+encodes a single categorical variable as a column for each category,
+with each row having False in all but one column (True).
+These are called dummy variables, or one-hot encoding.
```
Comment (jreback) [inline doc/source/user_guide/categorical.rst:138]: can u add a link to a wiki page (or scikit lean ok too)

File: doc/source/user_guide/categorical.rst [hunk 18]
```diff
@@ -129,6 +129,48 @@ This conversion is likewise done column by column:
...
+    cat
+
+    dummies = cat.to_dummies()
+    dummies
+
+    pd.Categorical.from_dummies(dummies)
+
+The :meth:`pandas.Categorical.from_dummies` class method accepts a dataframe
+whose dtypes are coercible to boolean, and an ``ordered`` argument
+for whether the resulting ``Categorical`` should be considered ordered
```
Comment (jreback) [inline doc/source/user_guide/categorical.rst:153]: you need to have some good cross links to the current get_dummies section

otherwise this is very confusing

i would prefer that these are actually in the get_dummies with just a small note here

File: pandas/core/arrays/categorical.py [hunk 19]
```diff
@@ -379,6 +382,137 @@ def __init__(
         self._dtype = self._dtype.update_dtype(dtype)
         self._codes = coerce_indexer_dtype(codes, dtype.categories)
 
+    @classmethod
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:385]: i agree with the others

either remove this or make it identical to get_dummies

File: pandas/core/arrays/categorical.py [hunk 20]
```diff
@@ -379,6 +382,137 @@ def __init__(
...
+                    multicat_rows.sum(), list(df.index[multicat_rows]),
+                )
+            )
+
+        mult_by = np.arange(df.shape[1]) + 1
+        #  000            000    0   -1
+        #  010            020    2    1
+        #  001 * 1,2,3 => 003 -> 3 -> 2 = correct codes
+        #  100            100    1    0
+        codes = ((df * mult_by).sum(axis=1, skipna=False) - 1).astype("Int64")
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:460]: this is actually a memory heavy impl

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py (hunk 19 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:385]: I think the discussion around this was more about the necessity/ implementation of `to_dummies` - there currently isn't an equivalent of `from_dummies` in pandas, which is what originally precipitated the issue. But note taken re. `to_dummies` if that was the intention.

File: pandas/core/arrays/categorical.py (hunk 20 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:460]: Sure, I guess the alternative would be something like

```python
codes = pd.Series(np.full(len(df), np.nan), dtype="Int64")

row_totals = df.sum(axis=1, skipna=False)
... # multicat_rows check

codes[row_totals == 0] = -1
row_idx, code = np.nonzero(row_totals > 0)
codes[row_idx] = code
```

**Comment by dsaxton** [comment]
@clbarnes Is this still active? Can you fix merge conflicts if so?

**Comment by clbarnes** [comment]
Sorry I haven't been able to get back to this, I'll see what I can do today.

**Comment by clbarnes** [comment]
I'm calling `pandas.get_dummies` in `Categorical.get_dummies`. I'd rather be calling `pandas.core.reshape.reshape._get_dummies_1d`, but that fails a lint, and I don't want to start renaming existing functions - any guidance on that?

**Comment by clbarnes** [comment]
Still to do:

- [ ] Edge cases for np.nonzero with "boolean" dtype in Categorical.from_dummies
- [ ] More tests, generally
- [ ] Check cohesiveness of docs

File: pandas/core/arrays/categorical.py [hunk 21]
```diff
@@ -370,6 +373,224 @@ def __init__(
...
+        3  0  0
+        4  0  0
+
+        >>> pd.Categorical(list('abc')).get_dummies(dtype=float)
+            a    b    c
+        0  1.0  0.0  0.0
+        1  0.0  1.0  0.0
+        2  0.0  0.0  1.0
+        """
+        # Would be better to use pandas.core.reshape.reshape._get_dummies_1d
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:580]: don't leave comments like this. why *cant* you use _get_dummies_1d ?

File: pandas/core/arrays/categorical.py [hunk 22]
```diff
@@ -370,6 +373,224 @@ def __init__(
...
+            )
+
+        codes = Series(np.full(len(row_totals), np.nan), index=df.index, dtype="Int64")
+        codes[row_totals == 0] = -1
+        row_idx, code = np.nonzero(df)
+        codes[row_idx] = code
+
+        return cls.from_codes(codes.fillna(-1), cats, ordered=ordered)
+
+    def get_dummies(
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:494]: pls type these

File: pandas/core/arrays/categorical.py [hunk 23]
```diff
@@ -370,6 +373,224 @@ def __init__(
...
+        ... )
+        >>> Categorical.from_dummies(multi)
+        Traceback (most recent call last):
+            ...
+        ValueError: 1 record(s) belongs to multiple categories: [0]
+        """
+        from pandas import Series
+
+        copied = False
+        to_drop = dummies.columns[isna(dummies.columns.values)]
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:454]: you don't need .values

File: pandas/core/arrays/categorical.py [hunk 24]
```diff
@@ -370,6 +373,224 @@ def __init__(
...
+            to_keep = []
+            for c in dummies.columns:
+                if isinstance(c, str) and c.startswith(pref):
+                    to_keep.append(c)
+                    cats.append(c[len(pref) :])
+            dummies = dummies[to_keep]
+
+        df = dummies.astype("boolean")
+        if fillna is not None:
+            df = df.fillna(fillna, inplace=copied)
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:473]: i would kill the copied flag as its not worth it and adding un-needed complexity.

File: pandas/core/arrays/categorical.py [hunk 25]
```diff
@@ -370,6 +373,224 @@ def __init__(
...
+
+        row_totals = df.sum(axis=1, skipna=False)
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:475]: why skipna?

File: pandas/core/arrays/categorical.py [hunk 26]
```diff
@@ -370,6 +373,224 @@ def __init__(
...
+        if row_totals.isna().any():
+            raise ValueError("Unhandled NA values in dummy array")
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:477]: is this tested?

File: pandas/core/arrays/categorical.py [hunk 27]
```diff
@@ -370,6 +373,224 @@ def __init__(
...
+
+        multicat_rows = row_totals > 1
+        if multicat_rows.any():
+            raise ValueError(
+                "{} record(s) belongs to multiple categories: {}".format(
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:482]: use f-strings, is this tested?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py (hunk 21 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:580]: Importing a function with a leading underscore fails a lint. It's one of the home-rolled lints, so it can't be ignored with `# noqa`.

File: pandas/core/arrays/categorical.py (hunk 24 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:473]: There was a previous comment about the memory footprint so I thought I'd prevent additional copies where possible, but I would also prefer the simplicity.

File: pandas/core/arrays/categorical.py (hunk 26 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:477]: There are some holes left in the tests, on the to do list

File: pandas/core/arrays/categorical.py (hunk 25 above)
Comment (clbarnes) [reply pandas/core/arrays/categorical.py:475]: If there is no explicit `fillna` policy given, and there are still NA values in the data, I'd prefer to raise an error rather than silently pass bunk data through. Therefore nans should not be skipped in this step so that they can be checked for in the next line.

**Comment by arw2019** [comment]
@clbarnes how's this going? If you're interested in continuing can you merge master & resolve conflicts and address comments?

**Comment by clbarnes** [comment]
Sure, will do - did the first 90% but things got busy before I could tackle the last 90%. I'll try to get to this next week.

**Comment by arw2019** [comment]
> Sure, will do - did the first 90% but things got busy before I could tackle the last 90%. I'll try to get to this next week.

Sounds good (& no rush ofc).  Ping us when you're ready for the next round of reviews

**Comment by jreback** [comment]
@clbarnes still active? I think this was pretty close.

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears to have gotten stale. Going to close for now but let us know if you can merge master and update and we'd be happy to reopen.
